enum.rb

```diff
@@ -0,0 +1,521 @@
+# frozen_string_literal: true
+
+#:nodoc:
+# rubocop: disable Metrics/ModuleLength
+module Enumerable
+  # *********************************
+  # Beginning of my_each method.
+  # *********************************
+  # rubocop: disable Metrics/AbcSize,Metrics/CyclomaticComplexity, Metrics/MethodLength, Metrics/PerceivedComplexity
+  def my_each(nums = 0)
+    checked = []
+    num = nums.is_a?(Integer) && nums >= 1 ? nums : 0
+    unless block_given?
+      (return to_enum :my_each, (print '=>', 'No block was given', "\n"))
+    end
+    (return nil, print('=>', 'Array is empty', "\n")) if empty?
+
+    0.upto(length - num) do |index|
+      if num.is_a?(Integer) && num >= 1
+        aux = []
+        index.upto(index + num - 1) do |i|
+          aux.push(yield self[i])
+        end
+        print aux, "\n" unless aux.empty?
+      else
+        checked.push(yield self[index]) unless self[index].nil?
+      end
+    end
+    print checked unless checked.empty?
+  end
+
+  # rubocop: enable Metrics/ModuleLength, Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/MethodLength, Metrics/PerceivedComplexity
+  # *********************************
+  # End of my_each method.
+  # *********************************
+
+  # *********************************
+  # Beginning of my_each_with_index method.
+  # *********************************
+  # rubocop: disable Metrics/AbcSize,Metrics/CyclomaticComplexity, Metrics/MethodLength, Metrics/PerceivedComplexity
+  def my_each_with_index(nums = 0)
```

Comment:
- [ ] should be identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] after calling the given block once for each element in self, returns the array itself
```ruby
my_each_output = ''
block = proc { |num, idx| my_each_output += "Num: #{num}, idx: #{idx}\n" }
array.my_each_with_index(&block) == array.each_with_index(&block) # should return true
```

Reply:
I run the required tests according to the definition of the Enumerable each method and tests performed satisfactorily. my_each method is similar to each method, according to the each method definition. It returns an Enumerable if no block and passes each element of the array into the block.